### PR TITLE
[REACTOS] Improve the format of our debug traces - CORE-12671

### PR DESCRIPTION
**Pull request by HBelusca**
We currently use the following debug trace format:

    <debug_class>:(<file>:<line>) <message>

(with "debug_class" being "trace", "warn", "err"), and, in the cases where
we call the macros "FIXME()", "TRACE()", "WARN()" or "ERR()" without
specifying the function name, like in:

    FIXME("(%params) message\n", params);

which is what Wine does and what we therefore get with code syncs, we
generate incomplete traces:

    fixme:(<file>:<line>) (params) message

that do not mention the name of the function involved. This is opposite
to what Wine shows:

    fixme:<module>:SomeFunc(params) message

and this forces us to introduce bad hacks, for instance see: d9a3398 .

This patch attempts to fix this by including the function name in our
debug traces, while keeping a format that is readily understandable by
Testman & al. without any changes needed:

    <debug_class>:(<file>:<line>):<channel>:func <message>

As a final addition, the patch also adds support for displaying the current
process ID in addition to the (already existing support for) thread ID.

### CODE REVIEW

**Comment by yagoulas** [comment]
Do you have an example of the output? Because I'm afraid we will end up with huge prefixes to the actual debug message.

**Comment by EricKohl** [comment]
I don't like the idea of adding the function name to each debug line. I would like to have the function name only in one place, right at the beginning of the executable code. That's why I think my change is not a  hack but a useful improvement! I think your suggested change is bloating the debug log instead!

**Comment by HBelusca** [comment]
You both have a good point. I can try to run this PR on our testbots & we will check how the debug log looks like.

**Comment by HeisSpiter** [comment]
I don't see the point of all this. Excepted perhaps changing things just for the sake of changing things. When I see lines like: "
fixme:(dll/win32/crypt32/msg.c:2443):crypt:CDecodeMsg_Update (0023F498, 004816C6, 17, 0): streamed update stub", my eyes just bleeds. There are way too many information on this line. I don't care about the channel, I can obviously understand what channel is being use from the filename. For function, I don't care either, I just need to look at the lines...
Basically, previous format was already enough.
For the record, making ReactOS print all this extra will make it even slower on debug builds.

**Comment by learn-more** [comment]
I'd rather see logging like:
`[level][Function]: Message`
than
`file(line): Message`

Especially in the case of UNIMPLEMENTED:

`file.c(123): Is unimplemented`
versus:
`[err][SomeFunction]: Is Unimplemented`

**Comment by HBelusca** [comment]
When I worked on Word 2010 I had to diff huge logs of msi.dll generated from reactos wrt. the ones from Wine, and having to add or remove the names of the functions in 500 mb logs is... how to say... a pain.
So the aim was to have them more similar in this regard.
Note that for non-members wine modules this should matter less.

**Comment by HBelusca** [comment]
An alternative would be to allow the "extended" debug lines to be enabled at runtime (via environment variable), or at compilation time, while keeping the default behaviour as it currently is (i.e. no channel+function name displayed).

**Comment by yagoulas** [comment]
If you end up using two modes depending on an environment variable, the one mode should look as much as possible to wine. If they don't print the file name we shouldn't do that either (given that this is a mode that can be activated only when we need it).

**Comment by gigaherz** [comment]
I would personally prefer a human-redable log to look like:
`Module: ERROR: Message @file:line`
or
`Module: Func(params): ERROR: Message @file:line`

So we may see in practice a log like: `shell32: SHDoSomething(2f47823,1,2,4): NOT IMPLEMENTED @c:/dev/reactos/dll/win32/shell32/whatever.c:32`

But it would also be great to have an option to have computer-readable output logs instead, for use in testbots.

**Comment by HBelusca** [comment]
The original point of this feature report / PR was to **minimally** extend the traces to make them more easily comparable with Wine (thus, having the debug-channel + function name added) **while at the same time** retaining compatibility with Testbot without changes on its side (by keeping the debug-level + filename and line number at the beginning).

Additional notes:

As you may have noticed, the debuglevel+filename/line is needed for testman: see this revert at https://svn.reactos.org/svn/reactos?view=revision&revision=62761
which disabled having the channel name + function line that had been introduced in r57257, see: https://svn.reactos.org/svn/reactos/trunk/reactos/lib/3rdparty/libwine/debug.c?r1=57257&r2=57256&pathrev=57257

Finally, note that in the current code we have in master (this, what has been there since r62761), the channel name + function name is only displayed when the trace functions are called without both any file name and line number.

**Comment by learn-more** [comment]
most wine logging assumes that function names are in the trace, f.e. this one is not really clear what is happening without the function name:

`trace:(R:\src\master\dll\win32\msi\dialog.c:694) dialog L"CustomizeDlg" event L"SelectionPath" control L"Description" attribute L"SelectionPath"`

**Comment by yagoulas** [comment]
I think it is worth implementing doig exactly what wine does (but only when a special env var is in place): fixme:<module>:SomeFunc(params) message

**Comment by HBelusca** [comment]
Ok so let's try to do this. Someone please suggest an env-var name for that?

**Comment by binarymaster** [comment]
> Someone please suggest an env-var name for that?

`DEBUGFORMAT=wine`

File: sdk/include/reactos/wine/debug.h [hunk 1]
```diff
@@ -93,7 +93,7 @@ struct __wine_debug_channel
        __WINE_DBG_LOG
 
 #define __WINE_DBG_LOG(args...) \
-    ros_dbg_log( __dbcl, __dbch, __RELFILE__, __FUNCTION__, __LINE__, args); } } while(0)
+    ros_dbg_log( __dbcl, __dbch, __RELFILE__, __LINE__, __FUNCTION__, args); } } while(0)
```
Comment (ThFabba) [inline sdk/include/reactos/wine/debug.h]: Why are you swapping these parameters by the way? That just adds noise to the diff.

Comment (HBelusca) [reply sdk/include/reactos/wine/debug.h]: For an aesthetical purpose only: you go from the file (containing many functions), to the function (coded on many lines), to the particular line.

**Comment by tkreuzer** [comment]
@HBelusca Can we close this?

**Comment by HBelusca** [comment]
Superseded by PR #112.
